### PR TITLE
Consolidate action form logic

### PR DESCRIPTION
**Pull request by kulyk**
Based on #29289 and #29290
There's also a follow up PR that just moves around utilities once again #29328

### Description

Action form logic ended up being spread across a few layers and components. This PR consolidates as much logic as possible inside the `ActionForm` component and a new `useActionForm` hook. That includes:

* cleaning up initial values
* generating form fields out of implicit action parameters
* stripping out unchanged values for implicit update actions
* cleaning up values before submitting a form
* figuring out submit button label props (variant and label are configurable)
* hiding fields that don't need user's input (parameters connected to dashboard filters)

Overall, these changes should make the `ActionForm` the go-to component to use when we need an action form.

You can see how easier it is to add a new action form by looking an changes to the `PublicAction` component:

<details>
<summary>Before</summary>

```tsx
function PublicAction({ action, publicId, onError }: Props) {
  const [isSubmitted, setSubmitted] = useState(false);
  const successMessage = getSuccessMessage(action);

  // Must manually handle implicit actions that don't have form fields
  // Has to be done every time the ActionForm is used
  const formSettings = useMemo(() => {
    const actionSettings = action.visualization_settings || {};
    const fieldSettings =
      actionSettings.fields ||
      generateFieldSettingsFromParameters(action.parameters);
    return {
      ...actionSettings,
      fields: fieldSettings,
    };
  }, [action]);

  const handleSubmit = useCallback(
    async (values: ParametersForActionExecution) => {
      try {
        // Must clean the values before submitting
        // Has to be done every time the ActionForm is used
        const parameters = setNumericValues(values, formSettings.fields);
        await PublicApi.executeAction({ uuid: publicId, parameters });
        setSubmitted(true);
      } catch (error) {
        onError(error as AppErrorDescriptor);
      }
    },
    [publicId, formSettings, onError],
  );

  if (isSubmitted) {
    return <FormResultMessage>{successMessage}</FormResultMessage>;
  }

  return (
    <FormContainer>
      <FormTitle>{action.name}</FormTitle>
      <ActionForm
        parameters={action.parameters}
        formSettings={formSettings}
        onSubmit={handleSubmit}
      />
    </FormContainer>
  );
}
```

</details>

<details>
<summary>After</summary>

```tsx
function PublicAction({ action, publicId, onError }: Props) {
  const [isSubmitted, setSubmitted] = useState(false);
  const successMessage = getSuccessMessage(action);

  const handleSubmit = useCallback(
    async (parameters: ParametersForActionExecution) => {
      try {
        await PublicApi.executeAction({ uuid: publicId, parameters });
        setSubmitted(true);
      } catch (error) {
        onError(error as AppErrorDescriptor);
      }
    },
    [publicId, onError],
  );

  if (isSubmitted) {
    return <FormResultMessage>{successMessage}</FormResultMessage>;
  }

  return (
    <FormContainer>
      <FormTitle>{getFormTitle(action)}</FormTitle>
      <ActionForm action={action} onSubmit={handleSubmit} />
    </FormContainer>
  );
}
```

</details>

**Changes to action/dashboard parameter mapping**

There's a common use-case for actions on dashboards when we don't want to show the full action form, but only parameters not mapped to dashboard filters.

For example, we can have an action doing `UPDATE subscriptions SET plan = {{ plan }} WHERE id = {{ subscription_id }}`. We could add this action to a dashboard called "Subscription Details" with a subscription ID filter and an object detail view on it. In that case, we don't really want people to manually specify the subscription ID, as we can take a value directly from the filter.

Prior to this change, we'd figure out which parameters are missing and pass them to the `ActionForm`. Once the `ActionForm` is submitted, the parent component receives the submitted values through the `onSubmit` prop, merge then with dashboard filter values and sends an HTTP request to run the action. Overall, it looked like that:

<details>
<summary>Snippet</summary>

```jsx
function DashboardActionForm({ action, dashboardFilterValues, runAction }) {
  const missingParameters = getNotProvidedParameters(action.parameters, dashboardFilterValues });

  const fieldSettings = action.visualization_settings.fields || generateFieldSettingsFromParams(action.parameters)

  const handleSubmit = values => {
    runAction({
      ...values,
      ...cleanValues(dashboardFilterValues),
    })
  }

  return (
    <ActionForm
       parameters={missingParameters}
       fieldSettings={fieldSettings}
       onSubmit={handleSubmit}
    />
  )
}
```

</details>

Given this design, the parent component had to ensure that dashboard filter values have expected types (e.g. coerse strings to numbers) which is one more thing an engineer had to keep in mind while working with action forms.

This PR solves this by changing the data flow and adding a `hiddenFields` prop to the `ActionForm`:

<details>
<summary>Snippet</summary>

```jsx
function DashboardActionForm({ action, dashboardFilterValues, runAction }) {
  const missingParameters = getNotProvidedParameters(action.parameters, dashboardFilterValues });
  const hiddenFields = missingParameters.map(parameter => parameter.id)

  return (
    <ActionForm
       action={action}

       // We pass dashboard filter values as initial values here
       // and also instruct the ActionForm to visually hide them.
       // In that way, they end up in the values object when the form is submitted,
       // but still remain hidden for the end user.
       // This approach should be much more reusable
       initialValues={dashboardFilterValues}
       hiddenFields={hiddenFields}

       // As the ActionForm now handles cleaning up values,
       // we no longer need to do this here
       onSubmit={runAction}
    />
  )
}
```

</details>

### How to verify

1. CI should be green
2. Play around with actions and ensure everything works the same as on `master`

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29326)

### CODE REVIEW

File: e2e/test/scenarios/models/model-actions.cy.spec.js
Comment (kulyk) [inline e2e/test/scenarios/models/model-actions.cy.spec.js:1]: These changes are most likely temporary as we've decided to use the action name as a default submit button label

File: frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx [hunk 1]
```diff
@@ -271,7 +271,7 @@ describe("Actions > ActionViz > Action", () => {
     it("should render as a button if no parameters are missing", async () => {
       await setup({
         settings: formSettings,
-        parameterValues: { "dash-param-1": "foo", "dash-param-2": "bar" },
+        parameterValues: { "dash-param-1": "foo", "dash-param-2": 2 },
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:274]: Action's `parameter_2` is [actually numeric](https://github.com/metabase/metabase/blob/978abfe65110ff788f447ec3ee90ba1dcc989234/frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx#L47), so the test started failing once we consolidated type checks and validation

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx [hunk 2]
```diff
@@ -21,32 +22,38 @@ import { shouldShowConfirmation } from "./utils";
...
   shouldDisplayButton,
   isEditingDashcard,
+  onSubmit,
 }: ActionFormProps) {
   const [showModal, setShowModal] = useState(false);
   const title = getFormTitle(action);
+  const { getCleanValues } = useActionForm({
+    action,
+    initialValues: dashcardParamValues,
+  });
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:56]: I have mixed feelings about adding `useActionForm` to `ActionVizForm`. Most of the time, it'd end up rendering the `ActionForm`, which also uses the hook.

The reason it's here (and the reason `useActionForm` exists at all, TBH) is to handle a corner case when all parameters are mapped to filters, and we just need to execute the action on a button click ([see line 66](https://github.com/metabase/metabase/blob/f9c9003ff12098b4096dbcac936759d420a85e69/frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx#L66)). I didn't want to add another utility for this single use-case as we're trying not to spread the logic all over the place

Would appreciate more sane alternatives

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx [hunk 3]
```diff
@@ -92,69 +97,32 @@ function ActionParametersInputForm({
...
   ]);
 
-  const fieldSettings = useMemo(
-    () =>
-      action.visualization_settings?.fields ??
-      // if there are no field settings, we generate them from the parameters and field metadata
-      generateFieldSettingsFromParameters(
-        missingParameters,
-        dashcard?.card?.result_metadata as unknown as Field[],
-      ),
```
Comment (kulyk) [inline frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:102]: FWIW, you wouldn't find this `result_metadata` twist repeated anywhere else again as we're no longer doing this on `master` (this is essentially a dead codepath)

File: frontend/src/metabase/actions/utils.ts [hunk 4]
```diff
@@ -254,19 +252,6 @@ export const getFormTitle = (action: WritebackAction): string => {
...
-) {
-  Object.entries(params).forEach(([key, value]) => {
-    if (fieldSettings[key]?.fieldType === "number" && !isEmpty(value)) {
-      params[key] = Number(value) ?? null;
-    }
-  });
-
-  return params;
-}
-
```
Comment (kulyk) [inline frontend/src/metabase/actions/utils.ts:269]: Removed in favor of [`formatSubmitValues`](https://github.com/metabase/metabase/blob/f9c9003ff12098b4096dbcac936759d420a85e69/frontend/src/metabase/actions/hooks/use-action-form/utils.ts#L33)

File: frontend/src/metabase/public/containers/PublicAction/PublicAction.unit.spec.tsx
Comment (kulyk) [inline frontend/src/metabase/public/containers/PublicAction/PublicAction.unit.spec.tsx:1]: These changes are most likely temporary as we've decided to use the action name as a default submit button label

**Review by ranquild — COMMENTED** [review]
Let's merge https://github.com/metabase/metabase/pull/29296 first

**Review by WiNloSt — APPROVED** [review]
TBH, it'd take more time to properly review this PR. Given the bug bash schedule today, I've only provided general comments, hoping at least it would make modifying/fixing (if any) any part of this PR easier.

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx [hunk 5]
```diff
@@ -9,79 +9,94 @@ import FormProvider from "metabase/core/components/FormProvider";
...
+    useActionForm({
+      action,
+      initialValues: rawInitialValues,
+    });
 
-  const form = useMemo(
-    () => getForm(parameters, formSettings?.fields),
-    [parameters, formSettings?.fields],
+  const editableFields = useMemo(
+    () => form.fields.filter(field => !hiddenFields.includes(field.name)),
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:61]: Sounds good.

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx [hunk 6]
```diff
@@ -72,24 +79,24 @@ function ActionVizForm({
     return (
       <>
         <ActionButtonView
-          onClick={onClick}
           settings={settings}
           isFullHeight={!isSettings}
           focus={isEditingDashcard}
+          onClick={onClick}
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:85]: I think we usually pass function props last, just an automatic reordering

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx [hunk 7]
```diff
@@ -19,69 +15,78 @@ import type {
...
+  }, [dashboard?.id, dashcard?.id, dashcardParamValues]);
 
   useEffect(() => {
     const hasValueFromDashboard = Object.keys(dashcardParamValues).length > 0;
     const canPrefetch = hasValueFromDashboard && dashboard && dashcard;
 
     if (shouldPrefetch) {
-      setPrefetchValues({});
+      setPrefetchedValues({});
       canPrefetch && fetchInitialValues();
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:90]: Good call. What do you think about keeping it like that for a bit tho? I wonder if @iethree remembers if there's a need to reset prefetched values even if we can't prefetch them

[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:90]: I'm fine with that.

[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:90]: > I wonder if @iethree remembers if there's a need to reset prefetched values even if we can't prefetch them

I don't recall specifically. I do recall that there was logic in here that handled inline forms on dashboards where a linked value would change and we would have to refetch prefetch data.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx [hunk 8]
```diff
@@ -16,27 +16,30 @@ import {
...
   dashcard: createMockDashboard({ id: 456 }),
-  onSubmit: jest.fn(() => ({ success: true })),
+  dashcardParamValues: {},
   onCancel: _.noop,
   onSubmitSuccess: _.noop,
+  onSubmit: jest.fn(() => ({ success: true })),
 };
 
-async function setup(options?: any) {
+function setup(options?: any) {
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx:42]: Fixed! Great catch, it immediately helped me find a few outdated props in tests, thank you! 🙏

**Review by ranquild — APPROVED** [review]

**Pull request merged**